File: 09-Graph/06-Cycle-Detection/cycle_detection_directed.c
#include <stdio.h>

#define MAX 100

int graph[MAX][MAX];
int visited[MAX];
int rec_stack[MAX];
int n;

int has_cycle(int node) {
    visited[node] = 1;
    rec_stack[node] = 1;

    for(int i = 0; i < n; i++) {
        if(graph[node][i]) {

            if(!visited[i]) {
                if(has_cycle(i))
                    return 1;
            }

            else if(rec_stack[i]) {
                return 1;
            }
        }
    }

    rec_stack[node] = 0;

    return 0;
}

int main() {
    int e;

    printf("Enter number of vertices: ");
    scanf("%d", &n);

    printf("Enter number of edges: ");
    scanf("%d", &e);

    printf("Enter directed edges (u v):\n");

    for(int i = 0; i < e; i++) {
        int u, v;

        scanf("%d %d", &u, &v);

        graph[u][v] = 1;
    }

    int cycle = 0;

    for(int i = 0; i < n; i++) {
        if(!visited[i]) {
            if(has_cycle(i)) {
                cycle = 1;
                break;
            }
        }
    }

    if(cycle)
        printf("Cycle detected!\n");
    else
        printf("No cycle detected.\n");

    return 0;
}

File: 09-Graph/06-Cycle-Detection/cycle_detection_directed.py
n = int(input("Enter number of vertices: "))
e = int(input("Enter number of edges: "))

graph = [[] for _ in range(n)]

print("Enter directed edges (u v):")

for _ in range(e):
    u, v = map(int, input().split())
    graph[u].append(v)

visited = [False] * n
rec_stack = [False] * n


def has_cycle(node):
    visited[node] = True
    rec_stack[node] = True

    for neighbor in graph[node]:

        if not visited[neighbor]:
            if has_cycle(neighbor):
                return True

        elif rec_stack[neighbor]:
            return True

    rec_stack[node] = False

    return False


cycle = False

for i in range(n):
    if not visited[i]:
        if has_cycle(i):
            cycle = True
            break

if cycle:
    print("Cycle detected!")
else:
    print("No cycle detected.")